### PR TITLE
[LoopPeel] Peel iterations based on and, or conditions

### PR DESCRIPTION
**Pull request by caojoshua**
For example, this allows us to peel this loop with a `and`:
```
for (int i = 0; i < N; ++i) {
  if (i % 2 == 0 && i < 3) // can peel based on || as well
    f1();
  f2();
```
into:
```
for (int i = 0; i < 3; ++i) { // peel three iterations
  if (i % 2 == 0)
    f1();
  f2();
}
for (int i = 3; i < N; ++i)
  f2();
```

### CODE REVIEW

**Comment by caojoshua** [comment]
original phabricator review https://reviews.llvm.org/D151403

File: llvm/lib/Transforms/Utils/LoopPeel.cpp [hunk 1]
```diff
@@ -351,11 +351,19 @@ static unsigned countToEliminateCompares(Loop &L, unsigned MaxPeelCount,
...
+  std::function<void(Value *)> ComputePeelCount =
+      [&](Value *Condition) -> void {
     if (!Condition->getType()->isIntegerTy())
       return;
 
     Value *LeftVal, *RightVal;
+    if (match(Condition, m_And(m_Value(LeftVal), m_Value(RightVal))) ||
+        match(Condition, m_Or(m_Value(LeftVal), m_Value(RightVal)))) {
+      ComputePeelCount(LeftVal);
+      ComputePeelCount(RightVal);
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/LoopPeel.cpp:363]: It would be safer to use a worklist instead of recursion here.

Comment (nikic) [reply llvm/lib/Transforms/Utils/LoopPeel.cpp:363]: What I had in mind here is stack overflow due to deep recursion. But yes, having a visited set is also a good idea, as it ensures there is no exponential blowup for some degenerate IR.

Comment (caojoshua) [reply llvm/lib/Transforms/Utils/LoopPeel.cpp]: I added a depth limit, and some new test cases.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**